doc/stream-telemetry/stream-telemetry-hld.md

```diff
@@ -0,0 +1,986 @@
+192-255: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 10"
+256-319: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+320-383: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 5"
+384-399: "Set ID = 256"
+400-415: "Set Length = 32 bytes"
+416-447: "observationTimeNanoseconds = 20000"
+448-511: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 15"
+512-575: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+576-639: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 6"
+640-655: "Set ID = 256"
+656-671: "Set Length = 32 bytes"
+672-703: "observationTimeNanoseconds = 30000"
+704-767: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 20"
+768-831: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+832-895: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 8"
+```
+
+### Bandwidth Estimation
+
+We estimate the bandwidth based only on the effective data size, not the actual data size. The extra information in a message, such as the IPFIX header (16 bytes), data prefix (4 bytes), and observation time milliseconds (4 bytes), is negligible. For example, if we want to collect 30 stats on 64 ports, and the chunk size is 100: $The Percentage Of Effective Data = \frac{8 \times 30 \times 64 \times 100_{Effective Data}}{16_{Header} + 4 \times 100_{Data Prefix} + 4 \times 100_{Observation Time Milliseconds} + 8 \times 30 \times 64 \times 100_{Effective Data}} \approx 99.9\%$ .
+
+The following table is telemetry bandwidth of one cluster
+
+| # of stats per port | # of ports per switch | # of switch | frequency (us) | Total BW per switch(Mbps) | Total BW(Mbps) |
+| ------------------- | --------------------- | ----------- | -------------- | ------------------------- | -------------- |
+| 30                  | 64                    | 10,000      | 10             | 12,288                    | 122,880,000    |
+
+- ${Total BW Per Switch} = \frac{{\verb|#| Of Stats Per Port} \times 8_{bytes} \times {\verb|#| Of Ports Per Switch} \times {Frequency} \times 1,000 \times 8}{1,000,000}$
+- ${Total BM} = {Total BW Per Switch} \times {\verb|#| Of Switch}$
+
+### Config DB
+
+Any configuration changes in the config DB will interrupt the existing session and initiate a new one.
+
+#### STREAM_TELEMETRY_PROFILE
+
+```
+STREAM_TELEMETRY_PROFILE|{{profile_name}}
+    "stream_state": {{enabled/disabled}}
+    "poll_interval": {{uint32}}
+    "chunk_size": {{uint32}} (OPTIONAL)
+    "cache_size": {{uint32}} (OPTIONAL)
+```
+
+```
+key                = STREAM_TELEMETRY_PROFILE:profile_name a string as the identifier of stream telemetry
+; field            = value
+stream_state       = enabled/disabled ; Enabled/Disabled stream.
+poll_interval      = uint32 ; The interval to poll counter, unit milliseconds.
+chunk_size         = uint32 ; number of stats groups in a telemetry message.
+cache_size         = uint32 ; number of chunks that can be cached.
+```
+
+#### STREAM_TELEMETRY_GROUP
+
+```
+STREAM_TELEMETRY_GROUP|{{profile_name}}|{{group_name}}
+    "object_names": {{list of object name}}
+    "object_counters": {{list of stats of object}}
```

Comment:
How do you plan to handle the update of the counter list or object list from the report POV? is it going to be a new template ID?

Reply:
Yes, I assume we will use a new template ID.
We treat any configuration update as a new profile. The previous session will be interrupted. The SAI report object will get a new SAI_TAM_REPORT_ATTR_REPORT_IPFIX_TEMPLATE_ID